src/Data/Macaw/CFGSlice.hs

```diff
@@ -0,0 +1,177 @@
+import qualified Data.Kind as Kind
+
+import qualified Data.Parameterized.Context as Ctx
+
+import           What4.Interface
+import           What4.Partial
+
+import           Lang.Crucible.Backend
+import qualified Lang.Crucible.CFG.Core as C
+import qualified Lang.Crucible.CFG.Expr as C
+import qualified Lang.Crucible.CFG.Reg as CR
+
+import qualified Lang.Crucible.Simulator as C
+
+import qualified Lang.Crucible.LLVM.MemModel as MM
+
+import qualified Lang.Crucible.Utils.MuxTree as C
+import qualified Data.Macaw.Symbolic as MS
+
+import qualified Data.Macaw.CFG as M
+import qualified Data.Macaw.Discovery.State as M
+import qualified Data.Macaw.Types as M
+
+import qualified Data.Macaw.Symbolic.Backend as MSB
+
+-- * Retaining original macaw block endings when slicing CFG
+
+-- | An enum corresponding to the kind of terminal statement that originally
+-- appeared at the end of a block.
+data MacawBlockEndCase =
+    MacawBlockEndJump
+  | MacawBlockEndCall
+  | MacawBlockEndReturn
+  | MacawBlockEndBranch
+  | MacawBlockEndFail
+  -- | An otherwise-unclassified arch exit. Some arch exits may end up being classified
+  -- as either calls or returns, according to 'archTermCase'
+  | MacawBlockEndArch
+  deriving (Eq, Enum, Bounded, Ord)
+
+
+class HasArchTermEndCase (f :: (M.Type -> Kind.Type) -> Kind.Type) where
+  archTermCase :: f v -> MacawBlockEndCase
+
+type HasArchEndCase arch = HasArchTermEndCase (M.ArchTermStmt arch)
+
+-- | A summary of a 'M.ParsedTermStmt', representing how the block ended and
+-- potentally the address to return to in the case of a function call.
+data MacawBlockEnd arch = MacawBlockEnd MacawBlockEndCase !(Maybe (M.ArchSegmentOff arch))
+
+-- | A crucible encoding of 'MacawBlockEnd', where the 'MacawBlockEndCase' as an 8-bit bitvector
+-- and the return address is a 'MM.LLVMPointerType'.
+type MacawBlockEndType arch = C.StructType (Ctx.EmptyCtx Ctx.::> C.BVType 8 Ctx.::> C.MaybeType (MM.LLVMPointerType (M.ArchAddrWidth arch)))
+
+blockEndAtom :: forall arch ids s
+              . MS.MacawSymbolicArchFunctions arch
+             -> MacawBlockEnd arch
+             -> MSB.CrucGen arch ids s (CR.Atom s (MacawBlockEndType arch))
+blockEndAtom archFns (MacawBlockEnd blendK mret) = MSB.crucGenArchConstraints archFns $ do
+    blendK' <- MSB.bvLit knownNat (toInteger $ fromEnum blendK)
+    let
+      memWidthNatRepr = M.memWidthNatRepr @(M.ArchAddrWidth arch)
+      ptrRepr = MM.LLVMPointerRepr memWidthNatRepr
+    mret' <- case mret of
+      Just addr -> do
+        ptr <- MSB.valueToCrucible $ M.RelocatableValue (M.addrWidthRepr (Proxy @(M.ArchAddrWidth arch))) (M.segoffAddr addr)
+        MSB.appAtom $ C.JustValue ptrRepr ptr
+      Nothing -> do
+        MSB.appAtom $ C.NothingValue ptrRepr
+    let repr = Ctx.empty Ctx.:> C.BVRepr knownNat Ctx.:> C.MaybeRepr ptrRepr
+    MSB.appAtom $ C.MkStruct repr (Ctx.empty Ctx.:> blendK' Ctx.:> mret')
+
+assignBlockEnd :: HasArchEndCase arch =>
```

Comment:
Can you just add some notes about what `assignBlockEnd` and `blockEndAtom` are working together to accomplish?